### PR TITLE
Translate errors

### PR DESCRIPTION
**Pull request by phoenixeliot**
Translates error messages via regex just before displaying them to the user. Also includes a page for checking how many of our common error messages are translated (which is a bit janky, but works)

### CODE REVIEW

File: app/locale/zh-HANS.coffee [hunk 1]
```diff
@@ -2279,7 +2279,7 @@ module.exports = nativeDescription: "简体中文", englishDescription: "Chinese
     reference_error: "参考错误："
     spelling_issues: "请注意拼写错误：你指的是`$1`而不是`$2`, 对吗？"
     capitalization_issues: "请注意大小写：`$1` 应该改成 `$2`？"
-    py_empty_block: "空变量。在$1的语句中，加入4个空格。"
+    # py_empty_block: "空变量。在$1的语句中，加入4个空格。" # {change}
```
Comment (nwinter) [inline app/locale/zh-HANS.coffee]: Why remove this translation?

File: spec/server/unit/locale.spec.coffee [hunk 2]
```diff
@@ -0,0 +1,91 @@
+english = require("../../../app/locale/en")
+langs = [
```
Comment (nwinter) [inline spec/server/unit/locale.spec.coffee]: Possible to do this without enumeration? It's already pretty error-prone to add new languages in the 3 places we have to make edits.

File: app/lib/aether/problems.coffee [hunk 3]
```diff
@@ -334,7 +334,6 @@ class HintCreator
     @context = context ? {}
 
   getHint: (code, {message, range, error, aether}) ->
-    console.log error
```
Comment (nwinter) [inline app/lib/aether/problems.coffee]: I think this log tends to be pretty helpful; perhaps the logging should be cleaned up, but would like to preserve it.

File: app/core/Router.coffee [hunk 4]
```diff
@@ -115,6 +115,8 @@ module.exports = class CocoRouter extends Backbone.Router
     'editor/thang-tasks': go('editor/ThangTasksView')
     'editor/verifier': go('editor/verifier/VerifierView')
     'editor/verifier/:levelID': go('editor/verifier/VerifierView')
+    'editor/i18n-verifier/:levelID': go('editor/verifier/i18nVerifierView')
+    'editor/i18n-verifier': go('editor/verifier/i18nVerifierView')
```
Comment (sderickson) [inline app/core/Router.coffee]: Should go over how this works, make sure it's clear. Is this something that'll be useful going forward, if so, what's the expected use case?

File: app/locale/zh-HANS.coffee (hunk 1 above)
Comment (sderickson) [inline app/locale/zh-HANS.coffee]: I think it doesn't include $2.

File: app/templates/editor/verifier/i18n-verifier-view.jade [hunk 5]
```diff
@@ -0,0 +1,107 @@
+.container
```
Comment (sderickson) [inline app/templates/editor/verifier/i18n-verifier-view.jade]: Add `// DNT` to this file, to mark as do not translate

File: app/views/editor/verifier/i18nVerifierView.coffee [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+    problems: ->
+      _.sortBy(_.flatten(Object.values(@problemsByLevel), true), (p) -> -p.count)
+  created: ->
+    i18n.setLng(@language)
+    @loadCampaigns()
+    @loadLanguage(@language).then =>
+      @setupRegexes()
+      @getProblems(@levelSlug).then (newProblems) =>
+        @compareStrings(newProblems)
+        @loading = false
```
Comment (sderickson) [inline app/views/editor/verifier/i18nVerifierView.coffee]: Remember, if you want we can use `co` in non-user-facing pages 😛

File: app/views/editor/verifier/i18nVerifierView.coffee [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    messageOrHint: ->
+      @compareStrings(@problems)
+  methods:
+    loadLanguage: (language) ->
+      new Promise (accept, reject) =>
+        loading = application.moduleLoader.loadLanguage(language)
+        if loading
+          application.moduleLoader.once 'load-complete', accept
+        else
+          accept()
```
Comment (sderickson) [inline app/views/editor/verifier/i18nVerifierView.coffee]: Perhaps change `loadLanguage` to return a promise by default? If it wouldn't take long to update the other places that function is used.

File: app/views/editor/verifier/i18nVerifierView.coffee [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    loadCampaigns: ->
+      $.get(
+        '/db/campaign',
+        (@campaigns) =>
+          @selectedCampaign = _.find(@campaigns, (c) -> c.name is "Dungeon")
+          for campaign in @campaigns
+            Vue.set(campaign, 'levelsArray', Object.values(campaign.levels))
+      )
```
Comment (sderickson) [inline app/views/editor/verifier/i18nVerifierView.coffee]: Add get campaigns function to `app/core/api`?

File: app/views/editor/verifier/i18nVerifierView.coffee [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+      else
+        return 'red'
+    getProblemsAndCompare: (levelSlug) ->
+      @getProblems(levelSlug).then (problems) =>
+        @compareStrings(problems)
+    getProblems: (levelSlug) ->
+      new Promise (accept, reject) =>
+        $.post(
+          '/db/user.code.problem/-/common_problems',
+          {startDay: @startDay, endDay: @endDay, slug: levelSlug},
```
Comment (sderickson) [inline app/views/editor/verifier/i18nVerifierView.coffee]: Same, add a function to `app/core/api` and you won't have to manually wrap it in a Promise.

File: app/views/editor/verifier/i18nVerifierView.coffee [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+        )
+    compareStrings: (problems) ->
+      $.i18n.setLng(@language)
+      problems.forEach (problem) =>
+        original = problem[@messageOrHint]
+        translated = Problem.prototype.translate(problem[@messageOrHint])
+        # distance = Levenshtein.get(_.last(original.split(':')), _.last(translated.split(':')))
+        # trimmed = original
+        # for regex in @regexes
+        #   trimmed = trimmed.replace(regex, '')
```
Comment (sderickson) [inline app/views/editor/verifier/i18nVerifierView.coffee]: Remove comments?

File: app/views/editor/verifier/i18nVerifierView.coffee [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+        for regex in @otherRegexes
+          if false and /argument.*has.*a.*problem/.test(original)# and /Target.*an.*enemy.*variable/.test(regex.toString())
+            console.log "===="
+            console.log regex
+            console.log trimmed
+            console.log trimmed.replace(regex, '').replace(/^\n/, '')
+            debugger if trimmed isnt trimmed.replace(regex, '').replace(/^\n/, '')
+          trimmed = trimmed.replace(regex, '').replace(/^\n/, '')
+        Vue.set(problem, 'translated', translated)
+        # Vue.set(problem, 'distance', distance)
```
Comment (sderickson) [inline app/views/editor/verifier/i18nVerifierView.coffee]: Also remove

File: app/views/editor/verifier/i18nVerifierView.coffee [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+        # for regex in @regexes
+        #   trimmed = trimmed.replace(regex, '')
+        trimmed = translated
+        for regex in @otherRegexes
+          if false and /argument.*has.*a.*problem/.test(original)# and /Target.*an.*enemy.*variable/.test(regex.toString())
+            console.log "===="
+            console.log regex
+            console.log trimmed
+            console.log trimmed.replace(regex, '').replace(/^\n/, '')
+            debugger if trimmed isnt trimmed.replace(regex, '').replace(/^\n/, '')
```
Comment (sderickson) [inline app/views/editor/verifier/i18nVerifierView.coffee]: Should this debugger still be here?

File: spec/server/unit/locale.spec.coffee (hunk 2 above)
Comment (sderickson) [inline spec/server/unit/locale.spec.coffee]: Agree

File: app/views/play/level/tome/Problem.coffee [hunk 13]
```diff
@@ -104,24 +104,59 @@ module.exports = class Problem
...
+  #   and replace them with capture groups (.+),
+  # returns a regex that will match against the error message
+  #   and capture any dynamic values in the text
+  makeTranslationRegex: (englishString) ->
+    escapeRegExp = (str) ->
+      # https://stackoverflow.com/questions/3446170/escape-string-for-use-in-javascript-regex
+      return str.replace(/[\-\[\]\/\{\}\(\)\*\+\?\.\\\^\$\|]/g, "\\$&")
+    new RegExp(escapeRegExp(englishString).replace(/\\\$\d/g, '(.+)').replace(/ +/g, ' +'))
 
   translate: (msg) ->
```
Comment (sderickson) [inline app/views/play/level/tome/Problem.coffee]: Can you add some tests to this? I'm thinking a handful of errors in a few languages. In order to get around that none of these languages are translated, you could use rot13 or insert some fake (google?) translated errors into the locales during test prep.

File: app/locale/zh-HANS.coffee (hunk 1 above)
Comment (phoenixeliot) [inline app/locale/zh-HANS.coffee]: It seems to have $1 where $2 should have been, and no $2 at all.

File: app/core/Router.coffee (hunk 4 above)
Comment (phoenixeliot) [inline app/core/Router.coffee]: Its only purpose is to show us how much of the top error messages for a given level(s) are translated into a given language. Mostly useful for making sure messages are included in `en.coffee` to be translated.

**Pull request closed without merging**